File: app/src/main/java/io/bhurling/privatebet/presenter/Presenter.java
package io.bhurling.privatebet.presenter;

import android.support.annotation.CallSuper;

import io.reactivex.disposables.CompositeDisposable;

public abstract class Presenter<V extends Presenter.View> {

    protected final CompositeDisposable disposables = new CompositeDisposable();

    protected V view;

    @CallSuper
    public void attachView(V view) {
        this.view = view;
    }

    @CallSuper
    public void detachView() {
        this.disposables.dispose();

        this.view = null;
    }

    public interface View {}
}

File: lib-arch/src/main/java/io/bhurling/privatebet/arch/Presenter.kt
package io.bhurling.privatebet.arch

import android.support.annotation.CallSuper

import io.reactivex.disposables.CompositeDisposable

abstract class Presenter<V : Presenter.View> {

    protected val disposables = CompositeDisposable()

    private var _view: V? = null

    protected val view: V
        get() {
            _view?.let { return it }
                    ?: error("Cannot access view before attachView() or after detachView()")
        }

    @CallSuper
    open fun attachView(view: V) {
        this._view = view
    }

    @CallSuper
    fun detachView() {
        this.disposables.dispose()

        this._view = null
    }

    interface View
}
